Firestore/core/test/firebase/firestore/testutil/testutil.h

```diff
@@ -51,7 +51,135 @@ namespace testutil {
  */
 constexpr const char* kDeleteSentinel = "<DELETE>";
 
-// Below are convenience methods for creating instances for tests.
+// Convenience methods for creating instances for tests.
+
+inline model::FieldValue Value(nullptr_t) {
+  return model::FieldValue::Null();
+}
+
+/**
+ * Creates a boolean FieldValue.
+ *
+ * @param bool_value A boolean value that disallows implicit conversions.
+ */
+template <typename T,
+          typename = typename std::enable_if<std::is_same<bool, T>{}>::type>
+inline model::FieldValue Value(T bool_value) {
+  return model::FieldValue::FromBoolean(bool_value);
+}
+
+// Overload that captures integer literals. Without this, int64_t and double
+// are equally applicable conversions.
+inline model::FieldValue Value(int value) {
+  return model::FieldValue::FromInteger(value);
+}
+
+inline model::FieldValue Value(int64_t value) {
+  return model::FieldValue::FromInteger(value);
+}
+
+inline model::FieldValue Value(double value) {
+  return model::FieldValue::FromDouble(value);
+}
+
+inline model::FieldValue Value(Timestamp value) {
+  return model::FieldValue::FromTimestamp(std::move(value));
+}
+
+inline model::FieldValue Value(const char* value) {
+  return model::FieldValue::FromString(value);
+}
+
+inline model::FieldValue Value(const std::string& value) {
+  return model::FieldValue::FromString(value);
+}
+
+inline model::FieldValue Value(const GeoPoint& value) {
+  return model::FieldValue::FromGeoPoint(value);
+}
+
+// This overload allows Object() to appear as a value (along with any explicitly
+// constructed FieldValues).
+inline model::FieldValue Value(const model::FieldValue& value) {
+  return value;
+}
+
+inline model::FieldValue ArrayValue(std::vector<model::FieldValue>&& value) {
+  return model::FieldValue::FromArray(std::move(value));
+}
+
+namespace details {
+
+/**
+ * Recursive base case for AddPairs, below. Returns the map.
+ */
+inline model::FieldValue::Map AddPairs(const model::FieldValue::Map& prior) {
+  return prior;
+}
+
+/**
+ * Inserts the given key-value pair into the map, and then recursively calls
+ * AddPairs to add any remaining arguments.
+ *
+ * @param prior A map into which the values should be inserted.
+ * @param key The key naming the field to insert.
+ * @param value A value to wrap with a call to Value(), above.
+ * @param rest Any remaining arguments
+ *
+ * @return The resulting map.
+ */
+template <typename ValueType, typename... Args>
+model::FieldValue::Map AddPairs(const model::FieldValue::Map& prior,
+                                const std::string& key,
+                                const ValueType& value,
+                                Args... rest) {
```

Comment:
It's more important to talk about the interface to `Object`--these are implementation details. I'll assume you really wanted to talk about that :-).

I started working on `MakeMap`/`AddPairs` because my first iteration had something like what you're describing:

```c++
model::FieldValue Object(std::initializer_list<std::pair<std::string, model::FieldValue>>);
```

You'd call it like `Object({{"foo", Object({{"bar", Value(50)}})}})`.

Costa proposed making `Object` a constructor of an object that's implicitly convertible to `FieldValue`, and this removes parens, but it's still a lot.

I wish C++ could bind key-value pairs with colons, as you do in javascript. That makes the binding clear without creating a mess of curly braces.

On balance I think I prefer having fewer curlies to signaling the pairwise binding. It doesn't hurt that this ends up looking like the java port too.

Reply:
sgtm.